### ISSUE TITLE
Heap use-after-free in mrb_gc_mark

### ISSUE THREAD

**Issue by clayton-shopify** [mruby/mruby#3674]
The following input produces a crash, when supplied to mirb:
```
class P < Array <
    def method_missing(a,*b)
        c = Array.new(128) {Array.new}
        1.times{
            Array.new{} 
            x
        }   
    end
end

m = Fiber.new do
x
end

GC.start

m.resume

def x()
x end x
x 
e
end
```
ASAN report:
```
==11194==ERROR: AddressSanitizer: heap-use-after-free on address 0x62f000212400 at pc 0x000108f8055f bp 0x7fff56ce92d0 sp 0x7fff56ce92c8
READ of size 4 at 0x62f000212400 thread T0
    #0 0x108f8055e in mrb_gc_mark gc.c:711
    #1 0x108f87737 in mark_context_stack gc.c:564
    #2 0x108f86b2f in mark_context gc.c:581
    #3 0x108f88ab3 in gc_mark_children gc.c:668
    #4 0x108f879ff in gc_gray_mark gc.c:892
    #5 0x108f85be6 in incremental_marking_phase gc.c:987
    #6 0x108f84f03 in incremental_gc gc.c:1091
    #7 0x108f808bc in incremental_gc_step gc.c:1126
    #8 0x108f7fab6 in mrb_incremental_gc gc.c:1170
    #9 0x108f7f3b8 in mrb_obj_alloc gc.c:509
    #10 0x108f404d1 in mrb_instance_alloc (mirb:x86_64+0x10003e4d1)
    #11 0x108f3fbec in mrb_instance_new (mirb:x86_64+0x10003dbec)
    #12 0x109071c26 in mrb_vm_exec (mirb:x86_64+0x10016fc26)
    #13 0x109066473 in mrb_vm_run (mirb:x86_64+0x100164473)
    #14 0x10905e30e in mrb_run (mirb:x86_64+0x10015c30e)
    #15 0x10905bc94 in mrb_funcall_with_block (mirb:x86_64+0x100159c94)
    #16 0x108f3fe89 in mrb_instance_new (mirb:x86_64+0x10003de89)
    #17 0x109071c26 in mrb_vm_exec (mirb:x86_64+0x10016fc26)
    #18 0x109066473 in mrb_vm_run (mirb:x86_64+0x100164473)
    #19 0x108f04359 in main mirb.c:549
    #20 0x7fffe5638234 in start (libdyld.dylib:x86_64+0x5234)

0x62f000212400 is located 49152 bytes inside of 49200-byte region [0x62f000206400,0x62f000212430)
freed by thread T0 here:
    #0 0x109318356 in wrap_free (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x56356)
    #1 0x108ff9db6 in mrb_default_allocf (mirb:x86_64+0x1000f7db6)
    #2 0x108f7be59 in mrb_free gc.c:270
    #3 0x108f8674f in incremental_sweep_phase gc.c:1059
    #4 0x108f84f4c in incremental_gc gc.c:1100
    #5 0x108f807a6 in incremental_gc_until gc.c:1116
    #6 0x108f80c1d in clear_all_old gc.c:1149
    #7 0x108f80117 in mrb_incremental_gc gc.c:1186
    #8 0x108f7f3b8 in mrb_obj_alloc gc.c:509
    #9 0x108fe9e22 in env_new (mirb:x86_64+0x1000e7e22)
    #10 0x108fe95ab in closure_setup (mirb:x86_64+0x1000e75ab)
    #11 0x108fe94c7 in mrb_closure_new (mirb:x86_64+0x1000e74c7)
    #12 0x109090f74 in mrb_vm_exec (mirb:x86_64+0x10018ef74)
    #13 0x109066473 in mrb_vm_run (mirb:x86_64+0x100164473)
    #14 0x108f04359 in main mirb.c:549
    #15 0x7fffe5638234 in start (libdyld.dylib:x86_64+0x5234)

previously allocated by thread T0 here:
    #0 0x109318520 in wrap_realloc (libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x56520)
    #1 0x108ff9d9f in mrb_default_allocf (mirb:x86_64+0x1000f7d9f)
    #2 0x108f7ab08 in mrb_realloc_simple gc.c:203
    #3 0x108f7b25e in mrb_realloc gc.c:217
    #4 0x108f7bce3 in mrb_malloc gc.c:238
    #5 0x108f7bd7d in mrb_calloc gc.c:256
    #6 0x108f7c329 in add_heap gc.c:326
    #7 0x108f7f417 in mrb_obj_alloc gc.c:512
    #8 0x108f404d1 in mrb_instance_alloc (mirb:x86_64+0x10003e4d1)
    #9 0x108f3fbec in mrb_instance_new (mirb:x86_64+0x10003dbec)
    #10 0x109071c26 in mrb_vm_exec (mirb:x86_64+0x10016fc26)
    #11 0x109066473 in mrb_vm_run (mirb:x86_64+0x100164473)
    #12 0x108f04359 in main mirb.c:549
    #13 0x7fffe5638234 in start (libdyld.dylib:x86_64+0x5234)

SUMMARY: AddressSanitizer: heap-use-after-free gc.c:711 in mrb_gc_mark
Shadow bytes around the buggy address:
  0x1c5e00042430: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c5e00042440: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c5e00042450: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c5e00042460: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x1c5e00042470: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x1c5e00042480:[fd]fd fd fd fd fd fa fa fa fa fa fa fa fa fa fa
  0x1c5e00042490: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c5e000424a0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c5e000424b0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c5e000424c0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c5e000424d0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==11194==ABORTING
Abort trap: 6
```
This issue was reported by https://hackerone.com/ahihi

**Comment by matz** [comment]
I think this issue was fixed by 650cfb9 along with #3675

**Comment by clayton-shopify** [comment]
@matz This input still produces a crash for me (in mirb only).

**Comment by matz** [comment]
Hmm, it works for me on Linux.  Do you use OSX?

**Comment by clayton-shopify** [comment]
Yes, I tested on OS X.

**Comment by matz** [comment]
I could reproduce the issue on my machine. It depends on compiler configuration. I haven't figured out the detail yet.  In any way, now I can investigate the issue.

**Comment by matz** [comment]
The last commit was wrong.

**Comment by matz** [comment]
The issue has already been solved, probably by b32ad130
